### ISSUE TITLE
Server Side Apply did not cleanup the managedFields

### ISSUE THREAD

**Issue by richardwxn** [kubernetes/kubernetes#96351]
**What happened**:
We recently integrate server side apply into Istio Operator. But we met some unexpected errors. For example, during istio installation, if the istio deployment is stuck in the ImagePullBackOff status for a bad hub specified, the istio service object in the cluster would endup with managedFields without fieldTypes: 
```
  managedFields:
  - apiVersion: v1
    manager: istio-operator
    operation: Apply
    time: "2020-11-08T22:30:01Z"
```
Then if we try to reapply the manifests into cluster with correct hub/tag, the patch request failed with this error `failed to decode managed fields: failed to convert managed fields from API: missing fieldsType in managed fields entry 0` 

**What you expected to happen**:
Either the existing service object should have a clean managedFields with fieldsType or the newly applied request would clean up the existing ones and proceed without failing hard. 

**How to reproduce it (as minimally and precisely as possible)**:
Install istio first with error, i.e. pointing to a non-existent hub/tag.
istioctl install --set 'WRONGHUB'
Then reinstalling with correct image hub set
istioctl install --set 'CORRECTHUB'

We expect the installation would go through, but instead it failed with error `failed to decode managed fields: failed to convert managed fields from API: missing fieldsType in managed fields entry 0`

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
`Client Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.5", GitCommit:"e0fccafd69541e3750d460ba0f9743b90336f24f", GitTreeState:"clean", BuildDate:"2020-04-16T11:44:03Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"17+", GitVersion:"v1.17.12-gke.1504", GitCommit:"17061f5bd4ee34f72c9281d49f94b4f3ac31ac25", GitTreeState:"clean", BuildDate:"2020-10-19T17:00:22Z", GoVersion:"go1.13.15b4", Compiler:"gc", Platform:"linux/amd64"}`
- Cloud provider or hardware configuration: GKE

**Comment by apelisse** [comment]
I'm unfamiliar with `istioctl`, what are the queries performed by `istioctl install --set 'WRONGHUB'`, we're really not supposed to end-up with an empty managedfields.

**Comment by apelisse** [comment]
@kwiesmueller Do you want to take a look at this?

**Comment by richardwxn** [comment]
@apelisse that command just helps to setting the hub for istio images in related deployment manifest, etc.

**Comment by kwiesmueller** [comment]
I'll take a look and try to reproduce it in a test.
/assign

**Comment by kwiesmueller** [comment]
@richardwxn could you point me to the place where you integrated SSA into your operator?

Also, can you give some examples as JSON or YAML of the patches you send that cause the corrupt managedFields?

Do you have a mutating webhook that might be responsible?

**Comment by richardwxn** [comment]
thx for taking a look, yep here is the code: https://github.com/istio/istio/blob/release-1.8/operator/pkg/helmreconciler/apply.go#L239.

I can send you example yamls during this process so you can see how the managedFields got corrupted.

Nope our mutatingwebhook does not look at the istio-system namespace in this case

**Comment by richardwxn** [comment]
And you can download the istioctl binary from here if you want to reproduce: https://gcsweb.istio.io/gcs/istio-prerelease/prerelease/1.8.0-alpha.3/

**Comment by richardwxn** [comment]
here is a output of a bad service resource: [bad-service.txt](https://github.com/kubernetes/kubernetes/files/5555365/bad-service.txt)

**Comment by kwiesmueller** [comment]
One note ahead that might not be directly related.
I don't think you can always rely on the feature being enabled just because the Kubernetes version is >1.16.
https://github.com/istio/istio/blob/release-1.8/operator/pkg/helmreconciler/apply.go#L94

I've looked through your apply code and I can't really see anything wrong at first sight.
Right now I don't have access to a fresh cluster to try it on, I'll see what I can do about that.

If you want to send me more yamls on slack, please do so.
It doesn't make much sense right now. Apply wouldn't leave an empty managedFields entry there and you can't update them.
The only explanation we could come up with so far is a mutating webhook as we might not be handling them correctly right now.

**Comment by kwiesmueller** [comment]
From the discussion on Slack it looks like there is a GKE specific Webhook that removes the managedFields/corrupts them.
This could be to an old client-go version in the mutating Webhook.

We should address the fact, that people can have their objects broken by mutating webhooks, even on apply.

**Comment by richardwxn** [comment]
The webhook is neg-annotation.config.common-webhooks.networking.gke.io, seems like it is fixed in k8s 1.18

**Comment by apelisse** [comment]
/close

This is a duplicate of #96688
